web rest api를 구현하라

tlatmsrud opened the pull request:
1. TaskController를 통한 기본 REST API 구현
2. NoDataException을 사용한 tasks 리스트의  노데이터 체크
3. Controller Advice기능을 제공하는 GlobalExceptionHandler를 사용한 404 응답

tlatmsrud committed:
web rest api를 구현하라

1. TaskController를 통한 기본 REST API 구현
2. NoDataException을 사용한 tasks 리스트의  노데이터 체크
3. Controller Advice기능을 제공하는 GlobalExceptionHandler를 사용한 404
   응답

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,90 @@
+
+    @GetMapping("/{id}")
+    public Task detail(@PathVariable("id") Long id){
+        Task task = findTask(id);
+
+        return task;
```
좌변에 변수가 아무것도 하는 일이 없는 것 같네요. 더 단순해지기도 하고, 변수로 추가적인 설명을 하지도 않고 그리고 반환된 값을 조작할 수도 없도록 바로 반환하는 것이 좋을 것 같습니다.

tlatmsrud replied:
넵 바로 반환하도록 수정하였습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
```
왜 NO_CONTENT로 응답해야 되는지 다음을 읽어보세요

> A successful response SHOULD be 200 (OK) if the response includes an
entity describing the status, 202 (Accepted) if the action has not
yet been enacted, or 204 (No Content) if the action has been enacted
but the response does not include an entity.

https://tools.ietf.org/html/rfc2616#section-9.7

tlatmsrud replied:
204 : 서버가 요청을 성공적으로 처리했지만 콘텐츠를 제공하지 않는다. 
현재 DELETE  API에는 정상 삭제처리될 경우에 응답 값이 없으니 204를 써야한다는 말씀이시군요! 감사합니다! 저는 단순히 204가 삭제 처리 완료와 같은 StatusCode인줄알았더니 ㅠㅠ 감사합니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,90 @@
+     */
+    private Task findTask(Long id){
+
+        Task findTask = tasks.stream()
+                .filter(task -> task.getId().equals(id))
+                .findFirst()
+                .orElse(null);
+
+        if(findTask == null){
+            throw new NoDataException();
+        }
```
Optional안의 값이 없으면 예외를 던지고 싶으셨군요. 다음과 같이 할 수 있습니다.

```java
        Task findTask = tasks.stream()
                .filter(task -> task.getId().equals(id))
                .findFirst()
                .orElseThrow(NoDataException::new);
```

혹은 Optional을 사용한다면 다음과 같이 사용하는게 더 일반적이긴 합니다.

```java
        Optional<Task> task = tasks.stream()
                .filter(task -> task.getId().equals(id))
                .findFirst()
        
        if (task.isPresent()) {
            // ... 
        }
```

tlatmsrud replied:
1번 방식으로 수정하였으며, 앞선 피드백과 마찬가지로 findTask 변수가 추가적으로 하는 일이 없어 바로 return하도록 수정하였습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    private Task generateTask(Task task){
+        rowId += 1;
```
한 번에 많은 요청이 왔을 때 어떻게 될지 고민해 보세요

tlatmsrud replied:
increaseTaskId();
task.setId(rowId);
tasks.add(task);
위 코드에서 increaseTaskId()에서 ID값에 +1 연산하는 시점부터 task.setId(rowId) 사이에 새로운 요청이 들어올 경우tasks.add(task) 작업을 하기 전 rowId를 한번 더 증가시킬 수 있을 것 같습니다.
그렇게 되면 동일한 ID를 가진 to-do item이 생길 수 있을 것으로 추측됩니다. 

찾아보니 메서드나 코드 단위로 쓰레드 세이프하도록 Lock, unLock을 처리해주는 synchronized 키워드가 있어 추가해봤습니다. 테스트는 thread.sleep 후 동일 API를 여러번 호출하는 방식으로 진행해봤습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,20 @@
+public class GlobalExceptionHandler {
+    @ExceptionHandler(NoDataException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public String handleEmailDuplicateException(NoDataException e){
+        return "data is null!";
+    }
```
사용자에게 data 가 null이라고 알려주고 있군요. 에러를 응답할 때는 클라이언트가 왜 이 오류가 났고 다음에는 어떻게 해야되는지 알려주는게 좋습니다. 예를들어서 이렇게 할 수 있을 것 같습니다.

```
중복된 이메일입니다. 이미 가입하셨는지 확인해 주세요.
```

위 에러는 왜 에러가 발생한 것이고 그래서 무엇을 해야하는지 알려주고 있습니다.

tlatmsrud replied:
존재하지 않는 ID 입니다. TO-DO List에 등록된 ID인지 확인해주세요. 로 변경해보았습니다.

tlatmsrud committed:
불필요한 코드는 제거하라

findTask 메서드 호출 결과를  task 변수에 담은 후 해당 변수를 return하고
있었으나, 하는 일이 없어 변수에 담을 필요 없이 바로 리턴하였다.

tlatmsrud committed:
orElse 대신 orElseThrow를 사용하라

orElse(null) 이후 null일 시 throw Exception을 하는 것 보다 orElseThrow를
사용하여 Optional의 value값이 null일 경우 throw Exception을 하는 것으로
수정하였다. 또한 리턴 값을 변수에 담을 필요가 없어 바로 리턴하였다.

tlatmsrud committed:
예외 메시지는 예외가 발생한 이유와 해결방안으로 구성해라

사용자가 예외 메시지를 통해 발생 이유와 해결방안을 알수 있도록 한다.

tlatmsrud committed:
여러 쓰레드가 공유해서 사용할 수 있는 값은 Thread Safe 한지 확인하라

taskId 값을 증가시키는 increaseTaskId에 대해 Thread Safe 하도록
synchronized 키워드를 추가하였다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,100 @@
+     *       }
+     *
+     *  2) 이를 람다식으로 표현하면 () ->new NoDataException() 로 나타낼 수 있다.
+     *  3) 이를 더블콜론으로 표현하면 NoDataException::new로 나타낼 수 있다.
```
https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html

위 문서를 읽어서 이번 기회에 메서드 레퍼런스에 대해서 자세히 배우고 가는 것도 좋습니다

tlatmsrud replied:
넵 감사합니다! 그런데 아직 공식문서 읽는게 익숙하지 않네요 ㅠㅠ

hannut91 replied:
읽기 어렵게 보여서 어려워보이지만, Java Tutorial은 정말 쉽게 써져 있습니다 ㅎㅎ 꼭 공식문서를 같이 보는 습관을 들여놓으세요

tlatmsrud replied:
넵 알겠습니다!
